### ISSUE TITLE
Add a maximum taxonomy size to the spec

### ISSUE THREAD

**Issue by chrisvls** [patcg-individual-drafts/topics#229]
The coarseness of the taxonomy is critical to the value proposition of the Topics API. Every explainer, summary, discussion, etc. that I can find describes it as such. 

- As written, the spec allows implementing browser to implement a taxonomy of 2,000,000,000 entries, consisting of the 2,000,000 most popular websites and 1,000 sub-sites within each. I believe that this would be well outside the intent of the project.

- As this is perhaps the key expression of the contract and compromise between advertising needs and user needs, it should be recorded in the spec.

- Without this limit defined in the spec, it is not reasonable to expect users to trust and understand the privacy implications of the feature.

- Without this limit defined in the spec, one can expect that it will drift upward over time. There are many considerations that will exert pressure to add more entries to the taxonomies. 

- Also, there are interesting questions about international and regional requirements. Without scarcity, these questions are probably not getting the vetting they need.

- The spec requires exact implementations of much less important behaviors – if the random topics are returned 7 percent or 10 percent or 4 percent of the time, it would not defeat the intent of the project. But this critical requirement is omitted.

I’d assert that:

- 500 is probably too small
- 50,000 is probably too large

Also:

- most explainers currently describe the taxonomy as under 2,000
- if the community thinks that 10,000 or 50,000 is coarse-grained, then the explainers are not setting expectations correctly, including in the TAG discussions

So I’d propose:

- In section 2, define topic id as an integer between 1 and 2,000.

Or, alternatively:

- In section 2, define topic id as an integer between 1 and 5,000.

**Comment by jkarlin** [comment]
My opinion here is that the specification should provide the algorithms and mechanisms for another user agent to be able to implement the API, but that the parameters themselves are implementer dependent. A browser may choose its own taxonomy. It could have a taxonomy of size 1 billion, but if it only ever returns 5 of those 1 billion, then there is no privacy harm. I think there should be non-normative text describing privacy concerns, to aid other user agents in their designs, but ultimately what parameters are chosen is up to them.

**Comment by martinthomson** [comment]
I might agree that the classifier could be independent, but there are requirements on its use that need to be specified and agreed.  For instance, I don't agree that a taxonomy can be arbitrarily large, because that is an essential parameter of the differential privacy  protections in the design.  Maybe you believe that it is within the rights of a browser maker to decide that privacy protection is not something their users will enjoy, but I won't agree to that and nor do I think that that is an acceptable position.  If this were to ever have a hope of being standardized, some minimum threshold for privacy would be a necessary component.

Also, implementing different taxonomies could have a significant effect on how the system is used.  So to avoid second-order effects on people who might end up using alternative taxonomies, it would be much better to have a single agreed and specified taxonomy.  For instance, if one browser never revealed a highly lucrative topic (Finance-related, say), some sites might (automatically or inadvertently, thanks ML) discriminate against the users of that browser.  In addition, given how there are certain biases inherent to taxonomy selection, if I were shipping this, I'd want some sort of oversight process with multiple stakeholders to ensure that revealing a particular value does not have consequences beyond was was intended.

**Comment by chrisvls** [comment]
So, I think a really important thing is that the spec match the explainer and the other assurances offered. If you think that increasing the taxonomy to a billion causes no privacy harm, please change the explainers and marketing so that you are ok living with the user expectations that you yourself are setting.

**Comment by chrisvls** [comment]
I didn't realize the W3C TAG discussion was active again. I have summarized some of my observations about the difference between the spec and the discussion [there](https://github.com/w3ctag/design-reviews/issues/726#issuecomment-1666066303).

**Comment by chrisvls** [comment]
My comments on the TAG discussion summarize some of the reasons why I am skeptical of the claim that a taxonomy of a billion items causes no privacy harms. I think there are two more that I will put here as they are more general:

- If the taxonomy is all the things promised (coarse-grained, cleansed of sensitive topics, human-curated and human-reviewable), then it is low stakes for the user if other protections fail. If the browser maker accidentally exposes the data through a breach or leak or bad implementation, if the sites figure out a way to collude, etc. There is no stronger security protection against a system leaking data than not putting the data in the system.

- It is easy to explain to users why they shouldn't worry about a coarse-grained taxonomy. I can explain it in a sentence. Explaining the chain of observable-caller, epoch rotation, topic randomization, etc. would be quite difficult. Explaining it in a sentence is probably impossible.

**Comment by dmarti** [comment]
@chrisvls It's still hard for me to understand what kind of limits on the size of taxonomy make sense, since it's possible for a caller to pass info about the user encoded as sets containing multiple topics. You might not consider an individual topic as sensitive, but a classifier on the caller's server can use it along with other topics to identify you as a member of a protected group. https://github.com/patcg-individual-drafts/topics/issues/221

**Comment by jkarlin** [comment]
> So, I think a really important thing is that the spec match the explainer and the other assurances offered.

Ah, gotcha. I don't think the spec is going to perfectly match the explainer in that I think we'd provide some room for implementer flexibility where possible. That said, Chrome's parameters _do_ match the explainer, and if it would be helpful we could publish Chrome's parameters in a separate document. And note that the Topics API does return the version name of the implementation it's using, so we've tried to make it easy for you to choose what to do based on the specific implementation version if you are so inclined.

When developing the API we ran studies to understand the reidentification rates of users based on the entire set of parameters. There is quite a bit of [research](https://arxiv.org/abs/2304.07210) behind the values we've chosen. So in that sense, I can see wanting to put in some reasonable constraints around those parameters. I do want to be careful not to overconstrain though. As I said before, the taxonomy could be quite large and still meet the reidentification standards by adjusting the other privacy parameters (e.g., increase the noise). 

In order to not enter reidentification analysis into the spec itself, perhaps we could settle on something fairly general and simple in practice instead. Something like,  "Browsers SHOULD choose the noise probability and taxonomy size so that at least 50 people per million will report any given topic in the taxonomy on a particular site and epoch. (This corresponds to 5% noise with a taxonomy of at most 1000 topics, approximately 2-3x the size of Chrome's v1 and v2 taxonomies.)"


> If you think that increasing the taxonomy to a billion causes no privacy harm, please change the explainers and marketing so that you are ok living with the user expectations that you yourself are setting.

I think there was poor communication here on my part. I was just trying to illustrate that an implementation could increase the taxonomy size and make up for it in some other way (never returning a billion minus five of the topics).  I'm sure you can come up with a much more realistic example.  My purpose was to show that we need to be careful not to overconstrain the specification, to allow room for innovative good ideas.

**Comment by chrisvls** [comment]
@dmarti You're right, picking a limit is tricky. That's why this is a very important part of the spec.

In the content management and enterprise application space, I have seen many features proposed where the solution would rely on a limited taxonomy. But, the features fail because it is not possible to craft a taxonomy that all user groups can agree on (the data users want it to be simple, the reporting users want too many). Or the taxonomy can be agreed, but it becomes almost immediately out-of-date. Or a limited taxonomy can be agreed in principle but the requirements of different languages, locales, business units can't converge.

By looking at whether a taxonomy limit can be agreed on now, we can determine if the whole feature will really work in practice. If we're not sure the limit os a solvable problem, then the feature will probably founder on one or more of these shores.

**Comment by chrisvls** [comment]
Interestingly, the Chrome UX does not, in fact, let the user review the taxonomy in its permissions UX, at least in non-GDPR areas. Oddly, it gives the user control to suppress sharing a topic, but perhaps only after Chrome has already shared that topic at least once. 

Setting that choice aside, @jkarlin it is still Google's intention to limit the size of the taxonomy for Chrome? If so, any response on why it might as well in the spec, given the current form of the spec and the arguments for the feature? If not, are there plans to update the TAG explainer, proposal, spec, etc? Is there any replacement for the privacy research offered so far?
